File: d03/d03.py
# take dx copies of iterator
# apply zip to them -> grouping each dx of them
# why does it not return just dx copies of each element?
# because iterator consumes each element, and we dont have deep copies
parts = lambda l, dx: zip(*[iter(l)]*dx)

# take common element and map it to score
# subtract 96: 'a' -> 1, 'A' -> -31
# send 'A' to 27 -> -31 % 58 == 27 -> apply mod 58
score = lambda part: (ord(set.intersection(*map(set, part)).pop())-96)%58

inp = open('input.txt').read().split()

print(sum(score(parts(l, len(l)//2)) for l in inp))
print(sum(map(score, parts(inp, 3))))

File: d03/d03.exs
half = &(div(length(&1), 2))

inp = File.read!("input.txt") 
      |> String.split("\n", trim: true)
      |> Enum.map(&to_charlist/1)

score = fn(inp) ->
        inp
        |> Enum.map(fn(s) -> 
          s
          |> Enum.map(&MapSet.new/1)
          |> Enum.reduce(&MapSet.intersection/2)
          |> Enum.random
          |> (&(rem(&1-?a+1 + 58, 58))).()
        end)
        |> Enum.sum
end

star1 = inp
        |> Enum.map(&(Enum.chunk_every(&1, half.(&1))))
        |> score.()

star2 = inp
        |> Enum.chunk_every(3)
        |> score.()

IO.inspect(star1)
IO.inspect(star2)
